lib/eig_trlm_3d.cpp

```diff
@@ -0,0 +1,701 @@
+                              int size)
+  {
+    if (size > (int)evecs.size())
+      errorQuda("Requesting %d eigenvectors with only storage allocated for %lu", size, evecs.size());
+    if (size > (int)evals.size())
+      errorQuda("Requesting %d eigenvalues with only storage allocated for %lu", size, evals.size());
+
+    ColorSpinorParam csParamClone(evecs[0]);
+    ColorSpinorField tmp(csParamClone);
+
+    std::vector<std::vector<Complex>> evals_t(size, std::vector<Complex>(ortho_dim_size, 0.0));
+    std::vector<double> norms(ortho_dim_size, 0.0);
+    std::vector<Complex> unit(ortho_dim_size, 1.0);
+    std::vector<Complex> n_unit(ortho_dim_size, {-1.0, 0.0});
+
+    for (int i = 0; i < size; i++) {
+      // r = A * v_i
+      mat(tmp, evecs[i]);
+
+      // lambda_i = v_i^dag A v_i / (v_i^dag * v_i)
+      // Block dot products stored in s_t.
+      blas3d::cDotProduct(evals_t[i], evecs[i], tmp);
+      blas3d::reDotProduct(norms, evecs[i], evecs[i]);
+      for (int t = 0; t < ortho_dim_size; t++) evals_t[i][t] /= sqrt(norms[t]);
+
+      // Measure ||lambda_i*v_i - A*v_i||
+      blas3d::caxpby(evals_t[i], evecs[i], n_unit, tmp);
+      blas3d::reDotProduct(norms, tmp, tmp);
+      for (int t = 0; t < ortho_dim_size; t++) residua_3D[t][i] = sqrt(norms[t]);
+    }
+
+    // If size = n_conv, this routine is called post sort
+    if (size == n_conv) {
+      evals.resize(ortho_dim_size * comm_dim_global(ortho_dim) * n_conv, 0.0);
+
+      if (comm_coord(0) == 0 && comm_coord(1) == 0 && comm_coord(2) == 0) {
+        int t_offset = ortho_dim_size * comm_coord_global(3);
+        for (int t = 0; t < ortho_dim_size; t++) {
+          for (int i = 0; i < size; i++) {
+
+            // Use printf to get data from t dim only
+            if (getVerbosity() >= QUDA_VERBOSE) {
+              printf("Eval[%02d][%04d] = (%+.16e,%+.16e) residual = %+.16e\n",t_offset + t, i,
+                     evals_t[i][t].real(), evals_t[i][t].imag(), residua_3D[t][i]);
+            }
+
+            // Transfer evals to eval array
+            evals[(t_offset + t) * size + i] = evals_t[i][t];
+          }
+        }
+      }
+    }
+  }
+
+  void TRLM3D::loadFromFile3D(const DiracMatrix &mat, std::vector<ColorSpinorField> &kSpace, std::vector<Complex> &evals)
+  {
+    // Set suggested parity of fields
+    const QudaParity mat_parity = impliedParityFromMatPC(mat.getMatPCType());
+    for (int i = 0; i < n_conv; i++) { kSpace[i].setSuggestedParity(mat_parity); }
+
+    // load the vectors
+    VectorIO io(eig_param->vec_infile, eig_param->io_parity_inflate == QUDA_BOOLEAN_TRUE);
+    io.load({kSpace.begin(), kSpace.begin() + n_conv});
+
+    // Error estimates (residua) given by ||A*vec - lambda*vec||
+    computeEvals3D(mat, kSpace, evals);
+  }
+
+  template <extremumType min_max, typename T>
+  T TRLM3D::getArrayMinMax3D(const std::vector<T> &array)
```

Comment:
I think it could be worth splitting up this routine and making it a bona-fide member of `comm_quda.h`, have something like `comm_allreduce_array_min` and `comm_allreduce_array_max` or something of the sort. This could have far more value than just this context here (though I can't think of exactly what for right now).

Reply:
We already have array min / max comm reduction functions, but that's not what's being used here.  What's here is simpler, and is a scalar reduction, not array reduction.  However, `getArrayMinMax3D` is completely unnecessary, and it's better to just delete it.  1ddda7c54 does this.